### DISCUSSION TITLE
[0.21] pass #[pyclass] by mutable reference to async fn

### DISCUSSION THREAD

**Discussion by rigens** [PyO3/pyo3#4047] [Questions] [answered]
I'm sorry, I'm new to pyo3. Can anyone help me how to pass `#[pyclass]` object to async function. With regular functions I usually use  `PyRefMut`, but it doesn't work with `async fn` because of `PyRefMut` is not `Send + 'static`

```rust
#[pyclass]
pub struct MyClass {}

#[pyfunction]
async fn foo(c: PyRefMut<MyClass>)->PyResult<()>{
    Ok(())
}
```

**Comment by adamreichold** [comment]
Take both `py: Python` and `c: Py<MyClass>` as parameters and call

```rust
let c = c.borrow_mut(py);
```

directly, i.e. call [`Py::borrow_mut`](https://docs.rs/pyo3/latest/pyo3/prelude/struct.Py.html#method.borrow_mut).

**Reply by rigens** [reply]
Unfortunately I have the same error:
```rust
#[pyclass]
pub struct MyClass {}

#[pyfunction]
async fn foo(py: Python<'_>, c: Py<MyClass>)->PyResult<()>{
    let c_ref = c.borrow_mut(py);
    Ok(())
}
```
```
error[E0277]: `*mut pyo3::Python<'static>` cannot be sent between threads safely
   --> src/lib.rs:73:1
    |
73  | #[pyfunction]
    | ^^^^^^^^^^^^^
    | |
    | `*mut pyo3::Python<'static>` cannot be sent between threads safely
    | within this `{async block@src/lib.rs:73:1: 73:14}`

```

**Reply by rigens** [reply]
@adamreichold ,  thanks,

with your help I found a working solution, but it doesn't seem very soundness to me

```rust
#[pyfunction]
async fn foo(c: Py<MyClass>)->PyResult<()>{
    Python::with_gil(|py|{
        let c_ref = c.try_borrow_mut(py)?;
        Ok(())
    })
}
```

**Reply by adamreichold** [reply]
Ok, you can work around this by calling [`Python::with_gil`](https://docs.rs/pyo3/latest/pyo3/marker/struct.Python.html#method.with_gil) yourself

```rust
#[pyclass]
pub struct MyClass {}

#[pyfunction]
async fn foo(c: Py<MyClass>)->PyResult<()>{
  Python::with_gil(|py| {
    let c_ref = c.borrow_mut(py);
    Ok(())
  })
}
```

but I wonder if this is a bug (the direct async integration is still pretty rough in 0.21) as I think I remember that we do hold the GIL when polling the future (with an opt-in to have it released before being polled).

**Reply by adamreichold** [reply]
> but it doesn't seem very soundness to me

It should certainly be sound, we aim to not provide any unsound safe API. But yes, it should not really be necessary IIRC.

**Comment by wyfo** [comment, marked as answer]
The error is caused by a [documented restriction](https://pyo3.rs/v0.21.0/async-await#send--static-constraint). 
The solution is also [documented](https://pyo3.rs/v0.21.0/async-await#implicit-gil-holding): you have to call `Python::with_gil` yourself, and it will have a negligible cost.

**Reply by adamreichold** [reply]
Understood. But shouldn't we carve out an exception for the GIL token just like we did for the receiver if we do hold the GIL during polling? If I understand things correctly, this is merely of questing of wiring things up correctly, not really of whether the token is `Send + 'static`?

**Reply by wyfo** [reply]
Indeed, I start thinking that for non `#[allow_threads]` function, we should be able to release the restriction, not only for the GIL token, but also for the bound arguments. The idea would be to capture the raw arguments (the `*mut #pyo3_path::ffi::PyObject` pointers, wrapping them to make them `Send`) and do the extraction inside the future, with `Python::assume_gil_acquired`. 

I may open a pull request after #3610 to do that, before #3611.

EDIT: Actually, I'm not sure the future will still be `Send` if we do that... ~~I need to try, and maybe, we will discuss about a unsafe but sound workaround by implementing `Send` ourself.~~

**Reply by davidhewitt** [reply]
The future can probably store `Py<T>` smart pointers which are `Send` and convert to `Bound<T>` at zero cost. All that would be needed is to have some trait wizardry to convert the types into `Py<T>` as an intermediate state.

**Reply by wyfo** [reply]
Ok, forget it. The future will not be `Send`, because you can't hold `!Send` values across `.await` points 🤦‍♂️  That's why this restriction exists, and cannot be dropped. I'm not even sure that `Ungil` auto trait will help, because `Python` should stay `!Send`.
We managed to hack the restriction for the receiver because `&self` is `Send`.

**Reply by wyfo** [reply]
@davidhewitt we will have the issue of holding `!Send` values across `.await` points.

**Reply by davidhewitt** [reply]
Ah good point 👍
